rfcs/001-mid-level-file-api.md

```diff
@@ -0,0 +1,194 @@
+# Mid-Level File Api
+
+## Summary
+
+This is a mid-level API wrapper around the various File related JavaScript apis. The possibility for a higher level API is left open.
+
+The mid-level file API aims to implement File I/O on top of the raw JS apis found here: https://w3c.github.io/FileAPI/ which includes listing files, creating files, and reading files.
+
+## The API
+
+First we have `FileList` which is simply a list of `File`s. This is an opaque struct that offers a way to iterate over `Files` and get individual files by index:
+
+```rust
+#[derive(Debug, Clone)]
+struct FileList { ... }
+
+impl FileList {
+  fn get(index: u32) -> Option<File> { ... }
+
+  fn len(&self) -> u32 { ... }
+
+  fn iter(&self) -> FileListIter { ... }
+
+  fn as_raw(&self) -> &web_sys::FileList { ... }
+}
+
+impl AsRef<web_sys::FileList> for FileList { ... }
+impl From<web_sys::FileList> for FileList { ... }
+impl Index<usize> for FileList {
+    type Output = File;
+    fn index(&self, index: usize) -> &Self::Output { ... }
+}
+```
+
+There is no `FileList::new` since creating a raw `web_sys::FileList` is not possible without going through a `web_sys::HtmlInputElement`.
+
+Next we the trait `BlobLike`
+```rust
+trait BlobLike {
+    fn size(&self) -> u64 { ... }
+
+    // the mime crate and mimes from blobs both conform to rfc6838
+    #[cfg(feature = "mime")]
+    fn mime_type(&self) -> Result<mime::Mime, mime::FromStrError> { ... }
+
+    fn raw_mime_type(&self) -> String { ... }
+
+    fn as_raw(&self) -> &web_sys::Blob;
+}
+```
+There are two structs that implement this trait: `Blob` and `File`.
+
+```rust
+#[derive(Debug, Clone)]
+struct Blob { ... }
+
+impl Blob {
+  fn new<T>(content: Option<T>) -> Blob
```

Comment:
I moved this to option, because I think it better represents that Blobs can be empty. We could require just `T` but then every time you'd want to create an empty blob, you'd have to do provide `&Vec::new()`. Maybe that's ok?

Reply:
There's a big difference between "empty" (which would be an empty `Vec`) and "nothingness" (which would be `None`).

Since `Blob` can be called with many different types (`Array`, `ArrayBuffer`, other `Blob`s, `String`, etc.) what would it mean for a blob to be "empty"? Would it be an empty string? An empty array?

Since it's mandatory to pass in an argument to the `Blob` constructor, I don't think using `Option` makes sense here.

Reply:
I agree. Ill make the change.